spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MongoExampleMapper.java

```diff
@@ -0,0 +1,243 @@
+import org.springframework.util.StringUtils;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBObject;
+
+/**
+ * @author Christoph Strobl
+ * @author Mark Paluch
+ * @since 1.8
+ */
+public class MongoExampleMapper {
+
+	private final MappingContext<? extends MongoPersistentEntity<?>, MongoPersistentProperty> mappingContext;
+	private final MongoConverter converter;
+	private final Map<StringMatcher, Type> stringMatcherPartMapping = new HashMap<StringMatcher, Type>();
+
+	public MongoExampleMapper(MongoConverter converter) {
+
+		this.converter = converter;
+		this.mappingContext = converter.getMappingContext();
+
+		stringMatcherPartMapping.put(StringMatcher.EXACT, Type.SIMPLE_PROPERTY);
+		stringMatcherPartMapping.put(StringMatcher.CONTAINING, Type.CONTAINING);
+		stringMatcherPartMapping.put(StringMatcher.STARTING, Type.STARTING_WITH);
+		stringMatcherPartMapping.put(StringMatcher.ENDING, Type.ENDING_WITH);
+		stringMatcherPartMapping.put(StringMatcher.REGEX, Type.REGEX);
+	}
+
+	/**
+	 * Returns the given {@link Example} as {@link DBObject} holding matching values extracted from
+	 * {@link Example#getProbe()}.
+	 *
+	 * @param example
+	 * @return
+	 * @since 1.8
+	 */
+	public DBObject getMappedExample(Example<?> example) {
+		return getMappedExample(example, mappingContext.getPersistentEntity(example.getProbeType()));
+	}
+
+	/**
+	 * Returns the given {@link Example} as {@link DBObject} holding matching values extracted from
+	 * {@link Example#getProbe()}.
+	 *
+	 * @param example
+	 * @param entity
+	 * @return
+	 * @since 1.8
+	 */
+	public DBObject getMappedExample(Example<?> example, MongoPersistentEntity<?> entity) {
+
+		DBObject reference = (DBObject) converter.convertToMongoType(example.getProbe());
+
+		if (entity.hasIdProperty() && entity.getIdentifierAccessor(example.getProbe()).getIdentifier() == null) {
+			reference.removeField(entity.getIdProperty().getFieldName());
+		}
+
+		ExampleSpecAccessor exampleSpecAccessor = new ExampleSpecAccessor(example.getExampleSpec());
+
+		applyPropertySpecs("", reference, example.getProbeType(), exampleSpecAccessor);
+
+		if (exampleSpecAccessor.isTyped()) {
+			Set<Class<?>> restrictedTypes = new HashSet<Class<?>>();
+			restrictedTypes.add(example.getResultType());
+			this.converter.getTypeMapper().writeTypeRestrictions(reference, restrictedTypes);
+		}
+
+		return ObjectUtils.nullSafeEquals(NullHandler.INCLUDE, exampleSpecAccessor.getNullHandler()) ? reference
+				: new BasicDBObject(SerializationUtils.flatMap(reference));
+	}
+
+	private String getMappedPropertyPath(String path, Class<?> probeType) {
+
+		MongoPersistentEntity<?> entity = mappingContext.getPersistentEntity(probeType);
+
+		Iterator<String> parts = Arrays.asList(path.split("\\.")).iterator();
+
+		final Stack<MongoPersistentProperty> stack = new Stack<MongoPersistentProperty>();
+
+		List<String> resultParts = new ArrayList<String>();
+
+		while (parts.hasNext()) {
```

Comment:
I'd love to see the nesting reduced here. Can we introduce eager returns/continues?